posts/2017-05-02-io-monad-for-cats.md

```diff
@@ -0,0 +1,404 @@
+
+This is precisely why `scala.concurrent.Future` is *not* a suitable type for encapsulating effects in this way: it isn't referentially transparent!  `Future` evaluates eagerly (sort of, see below) and memoizes its results, meaning that a `println` inside of a given `Future` will only evaluate *once*, even if the `Future` is sequenced multiple times.  This in turn means that `val x = Future(...); x; x` is not the same as `Future(...); Future(...)`, which is the very definition of a violation of referential transparency.
+
+Coming back to `IO`…  If `program` does not evaluate eagerly, then clearly there must be some mechanism for asking it to evaluate.  After all, Scala is not like Haskell: we don't return a value of type `IO[Unit]` from our `main` function.  `IO` provides an FFI of sorts for wrapping side-effecting code into pure `IO` values, so it must also provide an FFI for going in the opposite direction: taking a pure `IO` value and evaluating its constituent actions as side-effects.
+
+```scala
+program.unsafeRunSync()    // uh oh!
+```
+
+This function is called `unsafeRunSync()`.  Given an `IO[A]`, the `unsafeRunSync()` function will give you a value of type `A`.  You should only call this function *once*, ideally at the very end of your program!  (i.e. in your `main` function)  Just as with `IO.apply`, any expression involving `unsafeRunSync()` is not referentially transparent.  For example:
+
+```scala
+program.unsafeRunSync()
+program.unsafeRunSync()
+```
+
+The above will run `program` *twice*.  So clearly, referential transparency is out the window whenever we do this, and we cannot expect the normal laws and abstractions to remain sound in the presence of this function.
+
+### A sidebar on `Future`'s eager evaluation
+
+As Viktor Klang is fond of pointing out, `Future` doesn't *need* to evaluate eagerly.  It is possible to define an `ExecutionContext` in which `Future` defers its evaluation until some indefinitely later point.  However, this is not the default mode of operation for 99% of all `Future`s ever constructed; most people just use `ExecutionContext.global` and leave it at that.  Additionally, if someone hands me an arbitrary `Future`, perhaps as a return value from a function, I really have no idea whether or not that `Future` is secretly running without my consent.  In other words, the referential transparency (or lack thereof) of functions that I write using `Future` is dependent on the runtime configuration of some other function which is hidden from me.  That's not referential transparency anymore.  Because we cannot be *certain* that `Future` is deferring its evaluation, we must defensively assume that it is not.
+
+This, in a nutshell, is precisely why you should only use `Future` when interacting with legacy code or third-party APIs which require it, and when you do use `Future`, you should treat it as a side-effect just like `println` or `launchTheMissiles`.
```

Comment:
`Future` is an integral part of Scala's ecosystem and there are contexts in which `Future` is actually useful. When working in an imperative / dirty context, sometimes you really need `Promise` / `Future` references.

To give an example, in Monix I'm building a [AsyncSemaphore](https://github.com/monix/monix/blob/master/monix-execution/shared/src/main/scala/monix/execution/misc/AsyncSemaphore.scala) that uses Futures internally, because that was the best approach and then I'm wrapping that into a [TaskSemaphore](https://github.com/monix/monix/blob/master/monix-eval/shared/src/main/scala/monix/eval/TaskSemaphore.scala).

I would say this about `Future`: it's a low-level concurrency primitive and people should treat it as a low level concurrency primitive.